src/Orleans/Async/AsyncExecutorWithRetries.cs

```diff
@@ -114,75 +111,71 @@ internal static class AsyncExecutorWithRetries
             IBackoffProvider onSuccessBackOff = null,
             IBackoffProvider onErrorBackOff = null)
         {
-            if (maxExecutionTime != Constants.INFINITE_TIMESPAN && maxExecutionTime != default(TimeSpan))
-            {
-                DateTime now = DateTime.UtcNow;
-                if (now - startExecutionTime > maxExecutionTime)
-                {
-                    Exception timeoutException = new TimeoutException(String.Format("ExecuteWithRetries has exceeded its max execution time of {0}. Now is {1}, started at {2}, passed {3}",
-                            maxExecutionTime, LogFormatter.PrintDate(now), LogFormatter.PrintDate(startExecutionTime), now - startExecutionTime));
-                    throw timeoutException;
-                }
-            }
             T result = default(T);
-            int counter = callCounter;
-            Exception exception = null;
-            try
+            bool retry;
+
+            do
             {
-                callCounter++;
-                result = await function(counter);
+                retry = false;
 
-                bool retry = false;
-                if (callCounter < maxNumSuccessTries || maxNumSuccessTries == INFINITE_RETRIES) // -1 for infinite retries
+                if (maxExecutionTime != Constants.INFINITE_TIMESPAN && maxExecutionTime != default(TimeSpan))
                 {
-                    if (retryValueFilter != null)
-                        retry = retryValueFilter(result, counter);
+                    DateTime now = DateTime.UtcNow;
+                    if (now - startExecutionTime > maxExecutionTime)
+                    {
+                        throw new TimeoutException(
+                            $"ExecuteWithRetries has exceeded its max execution time of {maxExecutionTime}. Now is {LogFormatter.PrintDate(now)}, started at {LogFormatter.PrintDate(startExecutionTime)}, passed {now - startExecutionTime}");
+                    }
                 }
-                if (retry)
+
+                int counter = callCounter;
+
+                try
                 {
-                    if (onSuccessBackOff == null)
+                    callCounter++;
+                    result = await function(counter);
+
+                    if (callCounter < maxNumSuccessTries || maxNumSuccessTries == INFINITE_RETRIES) // -1 for infinite retries
                     {
-                        return await ExecuteWithRetriesHelper(function, callCounter, maxNumSuccessTries, maxNumErrorTries, maxExecutionTime, startExecutionTime, retryValueFilter, retryExceptionFilter, onSuccessBackOff, onErrorBackOff);
+                        if (retryValueFilter != null)
+                            retry = retryValueFilter(result, counter);
                     }
-                    else
+
+                    if (retry)
```

Comment:
This block of code could be moved to just before the `while`, allowing you to eliminate the similar block inside the `catch`.